rust/dev-pl011/src/lib.rs

```diff
@@ -0,0 +1,293 @@
+    if (isr & UART_TMIS_RXMIS) != 0 {
+        while (field!(dev, fr).read() & UART_TFR_RXFE) == 0 {
+            if CONSOLE_HAS_INPUT_BUFFER {
+                if uart.config.flag & PL011_FLAG_DEBUG_UART != 0 {
+                    let c = field!(dev, dr).read() as c_char;
+                    buf.write_char(c, false);
+                    continue;
+                }
+            }
+
+            // If we're out of rx buffer, mask the irq instead of handling it.
+            // TODO: This is racy, as the interrupt is owned by the user side code.
+            if buf.is_full() {
+                let old = field!(dev, imsc).read();
+                field!(dev, imsc).write(old & !UART_IMSC_RXIM);
+                break;
+            }
+            let c = field!(dev, dr).read() as c_char;
+            buf.write_char(c, false);
+        }
+
+        resched = true;
+    }
+
+    if resched {
+        sys::INT_RESCHEDULE
+    } else {
+        sys::INT_NO_RESCHEDULE
+    }
+}
+
+/// Early initialization sets the config so that panic print messages can output
+/// characters.
+#[unsafe(no_mangle)]
+extern "C" fn pl011_init_early(port: isize, config: *const pl011_config) {
+    PL011_UARTS[port as usize].early_init(|uart| unsafe {
+        (*uart).config = (*config).clone();
+    });
+}
+
+#[unsafe(no_mangle)]
+extern "C" fn pl011_init(port: isize) {
+    unsafe {
+        PL011_UARTS[port as usize].init(|uart| {
+            let uart = &mut *uart;
+            sys::cbuf_initialize(uart.rx_buf.get(), RXBUF_SIZE);
+
+            let mut dev = uart.get_dev();
+
+            sys::register_int_handler(
+                uart.config.irq,
+                Some(uart_irq),
+                uart as *mut Pl011Uart as *mut c_void,
+            );
+
+            // Clear all irqs
+            field!(dev, icr).write(0x3FF);
+
+            // Set fifo trigger level
+            field!(dev, ifls).write(0); // 1/8 full, 1/8 txfifo
+
+            // Enable rx interrupt
+            let old = field!(dev, imsc).read();
+            field!(dev, imsc).write(old | UART_IMSC_RXIM);
+
+            // Enable receive
+            let old = field!(dev, cr).read();
+            field!(dev, cr).write(old | UART_CR_RXEN);
+
+            // Enable interrupt
+            sys::unmask_interrupt(uart.config.irq);
+        })
+    };
+}
+
+#[unsafe(no_mangle)]
+extern "C" fn uart_putc(port: c_int, c: c_char) {
+    let uart = get_uart_early(port as usize);
+    let mut dev = uart.get_dev();
+
+    // Spin while fifo is full.
+    while (field!(dev, fr).read() & UART_TFR_TXFF) != 0 {}
+    field!(dev, dr).write(c as u32);
+}
+
+#[unsafe(no_mangle)]
+extern "C" fn uart_getc(port: c_int, wait: bool) -> c_int {
+    let uart = get_uart(port as usize);
+    let mut dev = uart.get_dev();
+    let buf = unsafe { Cbuf::new(uart.rx_buf.get()) };
+
+    if let Some(ch) = buf.read_char(wait) {
+        // TODO: This modify is racy, as the interrupt handler also does a
```

Comment:
Yah this will need to be wrapped in a spinlock here and in the IRQ to make that go away, I think we discussed something like that the other day.

It's okay for now though, this driver can go through some more clean up to work out a lot of details.